feat: add webhooks object selector support

eddycharly opened the pull request:
This PR adds object selector support to the webhooks.
It allows kyverno resources to easily exclude themselves by adding the `webhooks.kyverno.io/exclude` label.

This allow kyverno to rebirth in case of all pod crash at the same time.

To do:
- kustomize

To be discussed:
- label key
- combining this with namespace selector

MarcelMue commented on pkg/webhookconfig/registration.go:
```diff
@@ -220,28 +220,19 @@ func (wrc *Register) UpdateWebhookConfigurations(configHandler config.Interface)
-				continue
-			}
+		webhookCfg := config.WebhookConfig{}
+		if len(webhookCfgs) > 0 {
```
Should probably error and `continue` if it's not `>0`

eddycharly replied:
I think providing no config at all is valid, no ?

eddycharly replied:
But i wonder why it's an array... What's the point of accepting an array and taking only the first item ?

realshuting commented:
>It allows kyverno resources to easily exclude themselves by adding the webhooks.kyverno.io/exclude.

I love the idea of supporting `objectSelector` in the webhook. However, currently we cannot exclude Kyverno deployment as there's a mutating webhook `kyverno-verify-mutating-webhook-cfg` registered to perform the webhook health check. There's a WIP [enhancement](https://github.com/kyverno/kyverno/issues/2269#issuecomment-947721150) to move this check to use the Lease object, until then we need the admission requests for Kyverno deployments.

With this PR, can we still add support for configuring `objectSelector`, and make it consistent with the way we configure the `namespaceSelector` via the [ConfigMap](https://kyverno.io/docs/installation/#namespace-selectors)? This way we can load the configuration dynamically using the existing configmap informer. IMHO the annotation on the CRD itself is more of an informative message, and we should not use it for configuration.

This can be a very useful feature that allows granular control on the webhook. We can leverage it to exclude certain resources for default webhook configurations.
cc @JimBugwadia @chipzoller

eddycharly commented:
@realshuting I wondered what `kyverno-verify-mutating-webhook-cfg` was doing. Can you provide details about how it works ?
It looks like the handler does nothing ?

realshuting commented:
> @realshuting I wondered what `kyverno-verify-mutating-webhook-cfg` was doing. Can you provide details about how it works ?

Currently there's a timestamp annotation added to Kyverno deployment indicating the timestamp of the last seen admission request. This timestamp is updated whenever Kyverno receives the request in the webhook. However, if there's no activity happening on the cluster, Kyverno will trigger an update on the deployment itself to renew this timestamp, this way we can verify the webhook can actually receive requests thus healthy.


> It looks like the handler does nothing ?

Which handler?

eddycharly commented:
> Which handler?

The `verifyHandler` https://github.com/kyverno/kyverno/blob/main/pkg/webhooks/checker.go

eddycharly commented:
Ok, the important line is probably https://github.com/kyverno/kyverno/blob/cc10feb90651d3d6119be58331b033e002f9fc74/pkg/webhooks/server.go#L241

realshuting commented:
> Ok, the important line is probably
> 
> https://github.com/kyverno/kyverno/blob/cc10feb90651d3d6119be58331b033e002f9fc74/pkg/webhooks/server.go#L241

Yes, correct.

chipzoller commented:
I'm not sure I fully grasp this change.

eddycharly commented:
@chipzoller this allows excluding/including
 resources based on resource labels.
The first use case could be to exclude kyverno resources themselves.

realshuting commented:
@eddycharly - do you agree with my comment [above](https://github.com/kyverno/kyverno/pull/3413#issuecomment-1072277479)? What are your thoughts?

eddycharly commented:
@realshuting I'm not sure I understand your comment correctly, what shall we change to this PR ?
Also, I don't really understand what the verification process does. I see there is a logic around the annotation timestamp but what actions are taken when we detect something is going wrong ?

realshuting commented:
> @realshuting I'm not sure I understand your comment correctly, what shall we change to this PR ? Also, I don't really understand what the verification process does. I see there is a logic around the annotation timestamp but what actions are taken when we detect something is going wrong ?

My bad, let me try to clarify again. 

What I suggested is that:
- do not exclude Kyverno (at leaset Kyverno deployment) by default, as we use it for webhook checks
- update to use existing logic for `objectSelector` configuration (via ConfigMap), see doc for [namespaceSlector configuraiton](https://kyverno.io/docs/installation/#namespace-selectors)


>I see there is a logic around the annotation timestamp but what actions are taken when we detect something is going wrong ?

Kyverno will try to re-create webhook configurations if the webhook is detected unhealthy.

eddycharly commented:
> do not exclude Kyverno (at leaset Kyverno deployment) by default, as we use it for webhook checks

Ok, i'll remove the label on it.

> update to use existing logic for objectSelector configuration (via ConfigMap)

Isn't it the case already ? The changes in this PR read the `objectSelector` from the config map and updates the webhook accordingly.

realshuting commented on pkg/webhookconfig/registration.go:
```diff
@@ -746,18 +742,39 @@ func (wrc *Register) updateResourceValidatingWebhookConfiguration(nsSelector map
+			}
 
-		currentSelector, _, err := unstructured.NestedMap(webhook, "namespaceSelector")
-		if err != nil {
-			return errors.Wrapf(err, "unable to get validatingWebhookConfigurations.webhooks["+fmt.Sprint(i)+"].namespaceSelector")
-		}
-
-		if !reflect.DeepEqual(nsSelector, currentSelector) {
-			webhookChanged = true
+			if expectedSelector, err := convertLabelSelector(webhookCfg.ObjectSelector); err != nil {
+				return err
+			} else {
+				if !reflect.DeepEqual(expectedSelector, currentSelector) {
+					webhookChanged = true
+				}
+				if err = unstructured.SetNestedMap(webhook, expectedSelector, "objectSelector"); err != nil {
+					return errors.Wrapf(err, "unable to set validatingWebhookConfigurations.webhooks["+fmt.Sprint(i)+"].objectSelector")
+				}
+			}
```
Can we extract this logic to a separate method and re-use it for both selectors?

eddycharly replied:
Sure, I'll rework this.

eddycharly replied:
Done, created the `configureSelector` helper.

eddycharly commented on pkg/webhookconfig/registration.go:
```diff
@@ -712,7 +701,14 @@ func getHealthyPodsIP(pods []unstructured.Unstructured) (ips []string, errs []er
+		return nil, nil
+	}
+	return runtime.DefaultUnstructuredConverter.ToUnstructured(selector)
+}
```
Yes, I'm concerned about the memory issue. I'll need to investigate this deeper.

eddycharly commented:
@realshuting I think I addressed all comments.
I removed the label and logic entirely, this only adds support for `objectSelector` with labeling or ignoring anything special (will not disturb perturb webhook checks).

realshuting commented:
> @realshuting I think I addressed all comments. I removed the label and logic entirely, this only adds support for `objectSelector` with labeling or ignoring anything special (will not disturb perturb webhook checks).

Thank you @eddycharly !

With your change, we are able to filter out Kyverno resources particularly in the webhook, pointing to https://github.com/kyverno/kyverno/issues/3304.
